src/objects/key_value_store.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2024 the Cartero authors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+use glib::{object::Cast, subclass::types::ObjectSubclassIsExt, Object};
+use gtk::{gio, prelude::ListModelExt};
+
+use crate::entities::{KeyValue, KeyValueTable};
+
+use super::KeyValueItem;
+
+mod imp {
+    use std::cell::RefCell;
+
+    use glib::prelude::*;
+    use glib::subclass::prelude::*;
+    use gtk::gio::{self, ListStore};
+    use gtk::prelude::ListModelExt;
+    use gtk::subclass::prelude::ListModelImpl;
+
+    use crate::objects::KeyValueItem;
+
+    #[derive(Debug)]
+    pub struct KeyValueStore {
+        pub(super) inner: RefCell<ListStore>,
+    }
+
+    impl Default for KeyValueStore {
+        fn default() -> Self {
+            Self {
+                inner: ListStore::with_type(KeyValueItem::static_type()).into(),
+            }
+        }
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for KeyValueStore {
+        const NAME: &'static str = "CarteroKeyValueStore";
+        type Type = super::KeyValueStore;
+        type Interfaces = (gio::ListModel,);
+    }
+
+    impl ObjectImpl for KeyValueStore {
+        fn constructed(&self) {
+            let inner = self.inner.borrow();
+            let obj = &*self.obj();
+            inner.connect_items_changed(
+                glib::clone!(@weak obj => move |_, position, removed, added| {
+                    obj.emit_by_name::<()>("items-changed", &[&position, &removed, &added]);
+                }),
+            );
+        }
+    }
+
+    impl ListModelImpl for KeyValueStore {
+        fn item_type(&self) -> glib::Type {
+            let inner = self.inner.borrow();
+            inner.item_type()
+        }
+
+        fn n_items(&self) -> u32 {
+            let inner = self.inner.borrow();
+            inner.n_items()
+        }
+
+        fn item(&self, position: u32) -> Option<glib::Object> {
+            let inner = self.inner.borrow();
+            inner.item(position)
+        }
+    }
+}
+
+glib::wrapper! {
+    pub struct KeyValueStore(ObjectSubclass<imp::KeyValueStore>)
+        @implements gtk::Buildable, gio::ListModel;
+}
```

Comment:
Note to self: write integration tests to make sure that GTK is going to actually treat this object as reactive (for instance, when connecting the KeyValueStore into a KeyValuePane).